handlers/prometheus2_0.go

```diff
@@ -242,61 +248,72 @@ func PrometheusRead2_0(ctx echo.Context) error {
 
 		if err != nil {
 			ctx.Logger().Errorf("failed to find tags: %s", err.Error())
 			continue
 		}
 
-		for _, v := range tagResp {
-			// for all of our tag responses, grab the rollups
-			start := time.Now()
-			values, err := snowthClient.ReadRollupValues(
-				node, prp.checkUUID.String(), v.MetricName, []string{}, 60*time.Second,
-				time.Unix(0, q.StartTimestampMs*int64(time.Millisecond)),
-				time.Unix(0, q.EndTimestampMs*int64(time.Millisecond)),
-			)
+		ctx.Logger().Warnf("doing %d rollups for query: %s", len(tagResp), snowthTagQuery.String())
 
-			ctx.Logger().Debugf("rollup query: %s, result length: %d, duration: %v", v.MetricName, len(values), time.Now().Sub(start))
-			ctx.Logger().Debugf("rollup results: %+v", values)
-			if err != nil {
-				ctx.Logger().Errorf("failed to read rollup: %s", err.Error())
-				continue
-			}
+		var tsChan = make(chan *prompb.TimeSeries, 100)
 
-			timeSeries := new(prompb.TimeSeries)
-			labelPairs := []*prompb.Label{}
+		var step = 60 * time.Second
+		if q.Hints.StepMs > 0 {
+			step = time.Duration(q.Hints.StepMs) * time.Millisecond
+		}
 
-			for k, v := range responseTags {
-				labelPairs = append(labelPairs, &prompb.Label{
-					Name:  k,
-					Value: v,
-				})
-			}
-			timeSeries.Labels = labelPairs
-
-			for _, v := range values {
-				// convert value to time series
-				timeSeries.Samples = append(timeSeries.Samples,
-					&prompb.Sample{
-						Value:     v.Value,
-						Timestamp: v.Timestamp,
-					})
-			}
-			// add the timeseries to the query result
-			qr.Timeseries = append(qr.Timeseries, timeSeries)
+		for _, v := range tagResp {
+			go func() {
+				// for all of our tag responses, grab the rollups
+				start := time.Now()
+				values, err := snowthClient.ReadRollupValues(
+					node, prp.checkUUID.String(), v.MetricName, []string{}, step,
+					time.Unix(0, q.StartTimestampMs*int64(time.Millisecond)),
+					time.Unix(0, q.EndTimestampMs*int64(time.Millisecond)),
+				)
+
+				ctx.Logger().Warnf("timing rollup query: %s, result length: %d, duration: %v", v.MetricName, len(values), time.Now().Sub(start))
+				ctx.Logger().Debugf("rollup results: %+v", values)
+				timeSeries := new(prompb.TimeSeries)
+				if err != nil {
+					ctx.Logger().Errorf("failed to read rollup: %s", err.Error())
+					tsChan <- timeSeries
+					return
+				}
+
+				timeSeries.Labels = metricNameToLabelPairs(v.MetricName)
+
+				for _, v := range values {
+					// convert value to time series
+					timeSeries.Samples = append(timeSeries.Samples,
+						&prompb.Sample{
+							Value:     v.Value,
+							Timestamp: v.Timestamp * 1000, // prom does ms
+						})
+				}
+				// add the timeseries to the query result
+				tsChan <- timeSeries
+			}()
+		}
 
+		for range tagResp {
+			timeSeries := <-tsChan
+			ctx.Logger().Warnf("time series added to resultset, #samples: %d", len(timeSeries.Samples))
```

Comment:
I'm curious about this for loop.  If it should loop on all the results that come through tsChan, why not just make it something like:
`for timeSeries := range tsChan {`
Then the loop will automatically stop when tsChan closes.